parser-typechecker/src/Unison/Codebase/Branch.hs

```diff
@@ -211,80 +220,163 @@ branch0 terms types children edits =
+                  (\(r, n, t, v) -> (r, Name.fromReverseSegments (n NonEmpty.:| reversePrefix), t, v))
+                  (deepMetadataHelper (_types b0))
+          children <- deepChildrenHelper e
+          go (work <> children) (typeMetadata <> acc)
 
 -- | Derive the 'deepPaths' field of a branch.
-deriveDeepPaths :: Branch0 m -> Branch0 m
+deriveDeepPaths :: forall m. Branch0 m -> Branch0 m
 deriveDeepPaths branch =
-  branch {deepPaths = makeDeepPaths (nonEmptyChildren branch)}
+  branch {deepPaths = makeDeepPaths branch}
   where
-    makeDeepPaths :: Map NameSegment (Branch m) -> Set Path
-    makeDeepPaths children =
-      Set.mapMonotonic Path.singleton (Map.keysSet children) <> ifoldMap go children
+    makeDeepPaths :: Branch0 m -> Set Path
+    makeDeepPaths branch = State.evalState (go (Seq.singleton ([], 0, branch)) mempty) Set.empty
       where
-        go :: NameSegment -> Branch m -> Set Path
-        go n b =
-          Set.map (Path.cons n) (deepPaths $ head b)
+        go :: Seq (DeepChildAcc m) -> Set Path -> DeepState m (Set Path)
+        go Seq.Empty acc = pure acc
+        go (e@(reversePrefix, _, b0) Seq.:<| work) acc = do
+          let paths :: Set Path
+              paths =
+                if isEmpty0 b0
+                  then Set.empty
+                  else (Set.singleton . Path . Seq.fromList . reverse) reversePrefix
+          children <- deepChildrenHelper e
+          go (work <> children) (paths <> acc)
 
 -- | Derive the 'deepEdits' field of a branch.
-deriveDeepEdits :: Branch0 m -> Branch0 m
+deriveDeepEdits :: forall m. Branch0 m -> Branch0 m
 deriveDeepEdits branch =
-  branch {deepEdits = makeDeepEdits (_edits branch) (nonEmptyChildren branch)}
+  branch {deepEdits = makeDeepEdits branch}
   where
-    makeDeepEdits :: Map NameSegment (EditHash, m Patch) -> Map NameSegment (Branch m) -> Map Name EditHash
-    makeDeepEdits edits children =
-      Map.mapKeysMonotonic Name.fromSegment (Map.map fst edits) <> ifoldMap go children
+    makeDeepEdits :: Branch0 m -> Map Name EditHash
+    makeDeepEdits branch = State.evalState (go (Seq.singleton ([], 0, branch)) mempty) Set.empty
       where
-        go :: NameSegment -> Branch m -> Map Name EditHash
-        go n b =
-          Map.mapKeys (Name.cons n) (deepEdits $ head b)
+        go :: (Seq (DeepChildAcc m)) -> Map Name EditHash -> DeepState m (Map Name EditHash)
+        go Seq.Empty acc = pure acc
+        go (e@(reversePrefix, _, b0) Seq.:<| work) acc = do
+          let edits :: Map Name EditHash
+              edits =
+                Map.mapKeysMonotonic
+                  (Name.fromReverseSegments . (NonEmpty.:| reversePrefix))
+                  (fst <$> _edits b0)
+          children <- deepChildrenHelper e
+          go (work <> children) (edits <> acc)
+
+type DeepState m = State (Set (NamespaceHash m))
+
+type DeepChildAcc m = ([NameSegment], Int, Branch0 m)
+
+-- | Helper for knowing whether to descend into a child branch or not.
```

Comment:
Something I'm not totally understanding: `libDepth` tracking. I'm wondering whether the logic here allows one to use the `lib` name segment within their library (e.g. `lens.prisms.lib.whobaloo` or something, where `lens.lib.*` would be len's direct dependencies).

It seems to me that the (e.g.) terms we want to include in a branch's `deepTerms` are:

- shallow terms (i.e. this branch's terms that happen to have a 1 segment name)
- all non-`lib` childrens' deep terms (i.e. this branch's terms that happen to have a 2+ segment name)
- all `lib` childrens' shallow terms (i.e. direct dependencies' terms that happen to have a 1 segment name)
- all `lib` childrens' non-`lib` childrens' deep terms (i.e. direct dependencies' terms that happen to have a 2+ segment name)

And none of that requires any stateful tracking of how many `"lib"` we've seen.

Reply:
> I'm wondering whether the logic here allows one to use the lib name segment within their library

True, it doesn't; but we haven't come up with a better alternative

> It seems to me that the (e.g.) terms we want to include in a branch's deepTerms are

I didn't understand the alternative 😅 and also did think it still required tracking

Reply:
Ah, maybe I understand?

It's not captured in this thread, but we'd decided we want to include all the direct dependencies (<=1 lib) and at least one shallow-ish copy of transitive dependency names to allow `view` to continue to work at any depth without showing references for the time being.